DOCS-8072-4.3-tuning-app-design-kt (#1422)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/tuning-app-design.adoc

> = App Design Performance Best Practices

Reviewer: Something like...
```suggestion
= App Design That Maximizes Process Performance 
```
Your title is good, but it could be read as best practices for app design performance: the performance of design rather than the performance of the app

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> Consider the following best practices while you are designing your applications to enhance the performance. These practices covers various fronts such as the Mule runtime engine (Mule) version to use, what to avoid while designing your app, and specific settings when using Mule components and connectors.

Reviewer: ```suggestion
As you design your applications, consider the following best practices for enhancing processing performance. These practices include guidelines on choosing a Mule runtime engine (Mule) version, avoiding certain app design issues, and specific settings for using Mule components and connectors.
```

> == Mule Runtime Version

Reviewer: ```suggestion
== How to Choose the Best Mule Runtime Version for Your Application
```

Reviewer: Remember that the title in the left nav does not have to exactly match the wording of the title in the narrative (according to Mysti).

> Before you start tuning your application, first, check the following recommendations:

Reviewer: ```suggestion
Before you start tuning your application, review the following Mule version recommendations:
```

> * Use the latest Mule version to create and run the application.

Reviewer: ```suggestion
* Use the latest Mule version to create and run the application. +
This includes updating to the most recent patch of the version, if possible.
```

> * Update your application dependencies to the latest version so you take advantage of the recent performance improvements and bug fixes. Also, update to the latest version of the Mule patch whenever possible.

Reviewer: ```suggestion
* Update your application dependencies to the latest Mule version and patch to benefit from the most recent performance improvements and bug fixes. 
```

> * When developing and testing an application, use the same Mule version that will be used to deploy that application in your environments.

Reviewer: ```suggestion
* When developing and testing an application, use the same Mule version as used to deploy that application in your environments.
```

> == General App Design Recommendations

Reviewer: ```suggestion
== Tips to Design for Enhanced Processing Performance
```

> When you are designing your applications consider the following recommendations:

Reviewer: ```suggestion
As you design your apps, consider the following recommendations to increase performance:
```

> * Instead of using the Java module too many times to reuse some code (such as Java libraries), you can create an SDK module with operations that wrap Java invocations.

Reviewer: ```suggestion
* Instead of repeatedly using the Java module to reuse code such as Java libraries, create an SDK module with operations that wrap Java invocations.
```
"wrap" = "include"? "incorporate"? "loop"?

> * Use DataWeave to transform/enrich the data, and create payloads instead of custom scripting code (Java, Groovy, etc).

Reviewer: ```suggestion
* Use DataWeave to transform or enrich data, and create payloads instead of custom scripting code such as Java and Groovy.
```

> * Avoid sequences of `set-variable/set-payload` components. See the following example:

Reviewer: ```suggestion
* Avoid sequences of `set-variable/set-payload` operations such as in the following example:
```

[source,xml,linenums]
----
<set-variable value='"Hello"' doc:name="Set Variable" variableName="hello"/>
<set-variable value='"world!"' doc:name="Set Variable" variableName="world"/>
----

> Instead, squash these components into a single `ee:transform` that avoids generating a new event for each component. See the following example:

Reviewer: ```suggestion
Instead, combine them into a single `ee:transform` operation that avoids generating a new event for each component:
```

[source,xml,linenums]
----
<ee:transform doc:name="Transform Message" >
	<ee:variables >
		<ee:set-variable variableName="hello" ><![CDATA[%dw 2.0
            output application/json
            ---
           "Hello"]]></ee:set-variable>

           <ee:set-variable variableName="world" ><![CDATA[%dw 2.0
            output application/json
            ---
           "world"]]></ee:set-variable>
	</ee:variables>
</ee:transform>
----

> * Avoid store transient values to be used in `set-variable`. See the following example:

Reviewer: ```suggestion
* Avoid using store transient values in `set-variable`, as shown in the following example:
```

[source,xml,linenums]
----
<flow name="FileComponentWithsetVariableFlow" doc:id="daa9e38a-bb8c-4c87-a4c0-ba2dd90316b2" >
       <set-variable value="#['file_path/' ++ attributes.queryParams['filename']]" doc:name="Set Variable" doc:id="cf876447-12a6-4267-9b26-5e849153142c" variableName="fileName"/>
       <file:read doc:name="Read" doc:id="e4ad1ad4-0edb-45c9-ab43-d7f07effa277" path="#[vars.fileName]"/>
</flow>
----

> Instead, inline these values into components that need them. See the following example:

Reviewer: ```suggestion
Instead, move these values into components that need them:
```

[source,xml,linenums]
----
<flow name="FileComponentWithoutsetVariableFlow" doc:id="0d44d155-cabe-4a79-99e4-81de22743edb" >
      <file:read doc:name="Read" doc:id="09dfb229-67fd-4508-862f-489c123527ba" path="#['file_path/' ++ attributes.queryParams['filename']]"/>
</flow>
----

> * Avoid creating property resolvers, which causes the usage DataWeave for every event, applying an overhead. See the following example:

Reviewer: ```suggestion
* Avoid creating property resolvers, which require using DataWeave for every event and thereby increase process overhead:
```

[source,xml,linenums]
----
<flow-ref name="#[p(someProp) default p(otherProp)]"/>
----

> Instead, place DataWeave logic in a property placeholder and use a static flow reference. See the following example:

Reviewer: ```suggestion
Instead, place DataWeave logic in a property placeholder and use a static flow reference:
```

[source,xml,linenums]
----
<flow-ref name="${propertyContainer.propertyName}"/>
----

> == Components and Connectors

Reviewer: ```suggestion
== Components and Connectors That Impact Performance
```

When designing your apps using Mule components and connectors, consider the following recommendations to enhance the performance.

=== APIkit

> To obtain performance improvement, you can disable APIkit validations. In case payload validations are required, APIkit  validates input requests against API specification.

Reviewer: ```suggestion
To improve performance, you disable APIkit validations. It payload validations are required, APIkit validates against the API specification.
```

=== DataWeave

> When using DataWeave consider the following:

Reviewer: ```suggestion
When using DataWeave, consider the following ways to improve performance:
```

> * When dealing with large payloads, include the directive `indent=false` to improve the client's parsing performance and to reduce the response payload size.

Reviewer: ```suggestion
* If payloads are large, include `indent=false` to improve client parsing and reduce response size.
```

> * Define the input `content-type` to avoid verbose messages in the Log.

Reviewer: ```suggestion
* Define `content-type` to avoid verbose messages in the log.
```

> * Use inline DataWeave scripts on the development phase, then pass it to an external file to have a clean XML and reusable scripts.

Reviewer: ```suggestion
* Use inline DataWeave scripts during development and then move them to an external file to both clean app XML and retain scripts for reuse.
```

* Do not log complex DataWeave expressions with every request.

> * Avoid patterns of iterating over a collection and then transforming each individual item. See the following example:

Reviewer: ```suggestion
* Avoid patterns of iterating over a collection and then transforming each individual item:
```

[source,xml,linenums]
----
<foreach>
	<ee:transform >
		<ee:message >
			<ee:set-payload ><![CDATA[%dw 2.0
output application/json
---
payload mapObject (value, key) -> {
	(upper(key)): value
}
			]]></ee:set-payload>
		</ee:message>
	</ee:transform>
	...
</foreach>
----

> Instead, transform the entire collection and iterate later. See the following example:

Reviewer: ```suggestion
Instead, transform the entire collection and iterate later:
```

[source,xml,linenums]
----
<ee:transform>
	<ee:message >
		<ee:set-payload ><![CDATA[%dw 2.0
output application/json
---
payload map (item, index) -> {
	book: item mapObject (value, key) -> {
		(upper(key)): value
	}
}
		]]></ee:set-payload>
	</ee:message>
</ee:transform>
<foreach>
	...
</foreach>
----

=== Validator

> Use Validator instead of throwing exceptions with Groovy Scripts.

Reviewer: Use Validator to avoid throwing exceptions with Groovy scripts?

> ===  HTTP(s)

Reviewer: ```suggestion
===  HTTP and HTTPS
```

> For HTTP(s) consider the following recommendations:

Reviewer: ```suggestion
When using HTTP or HTTPS, consider the following recommendations:
```

> * Use the authentication mechanisms provided by Anypoint Connector for HTTP (HTTP Connector) instead of doing it manually (for example, manually generating the Basic auth header with a DataWeave expression).

Reviewer: ```suggestion
* Use the authentication mechanisms provided by Anypoint Connector for HTTP (HTTP Connector) instead using manual authentications such as manually generating the Basic auth header with a DataWeave expression.
```

> * If HTTPS is required, use the HTTPS Listener with a TLS Context reference, configuring the key-store with key-pair values generated with Keytool. The HTTPS configuration can be avoided if the SSL can also be terminated at the Load Balancer level when using a dedicated load balancer in CloudHub or if there is an on-premises load balancer for hybrid scenarios.

Reviewer: ```suggestion
* If HTTPS is required, use the HTTPS Listener with a TLS Context reference, configuring the keystore with key-pair values generated with Keytool. +
You can avoid HTTPS configuration if the SSL can also be terminated at the load-balancer level when using a dedicated load balancer in CloudHub or if there is an on-premises load balancer for hybrid scenarios:
```

> See the following example:

Reviewer: omit
```suggestion
```

[source,xml,linenums]
----
<tls:context name="TLS_Context" doc:name="TLS Context">
      <tls:key-store type="jks" path="company-keystore.jks" alias="${jks.alias}" keyPassword="${jks.key.password}" password="${jks.password}"/>
</tls:context>
----

=== Database Connector

> Use parameterized queries. Use dynamic queries only if you don't have another choice, remember to validate inputs previously. Using dynamic queries could cause SQL-Injection vulnerability.

Reviewer: ```suggestion
To avoid SQL-injection vulnerability, use parameterized queries. 
If you must use dynamic queries, remember to validate inputs first.
```

=== File Connector

> Instead of using a transform component for the transformation in the write operation, use a DataWeave expression. This avoids the message payload changes for the operations after the write. See the following example:

Reviewer: This avoids message payload changes that occur after the write operation?

Reviewer: ...after the write operation, as shown in the following example:

[source,xml,linenums]
----
<file:write path="output.csv">
 <file:content>#[%dw 2.0
                 output application/csv
                 ---
                 payload.customers.email
               ]
 </file:content>
</file:write>
----

> Use the recursive parameter to list files and folders within subfolders. See the following example:

Reviewer: ```suggestion
Use the recursive parameter to list files and folders within subfolders:
```

[source,xml,linenums]
----
<file:list config-ref="File_Config" directoryPath="relativePath" recursive="true" />
----

=== Web Service Consumer Connector

> Import WSDLs including all references (XSDs) in your project, use Anypoint Connector for Web Service Consumer (Web Service Consumer Connector) and point the configuration to the local WSDL.

Reviewer: ```suggestion
To import WSDLs including all references (XSDs) in your project, use Anypoint Connector for Web Service Consumer (Web Service Consumer Connector) and point the configuration to the local WSDL.
```

=== AMQP Connector

> The server constrains the performance and configuration of the connector. Therefore, when you are investigating performance issues, you must explore the server-side configuration to determine if consume or publish parameters have an impact on the performance.

Reviewer: ```suggestion
Because the server constrains the performance and configuration of Anypoint Connector for AMQP, when you investigate performance issues, explore the server-side configuration to determine if consume or publish parameters have an impact on performance.
```

==== Consume Parameters

> Review consume parameters that can have an impact on performance.

Reviewer: Why isn't this introduction to the subtopics as robust as the one for Publish Parameters?

> ===== Number of Consumers

Reviewer: Instead of using Level-4 and Level-5 subheads, can we use bullets and subbullets?
Hmm...I guess that would result in subsubbullets. Well darn.

> You can set the number of consumers for an AMQP configuration and the Listener source. It manages the number of consumers spawned by the source to receive AMQP messages concurrently, as each consumer creates and uses a different channel. +

Reviewer: To what does "It" refer? the number? That would then be "The number of consumers manages the number of consumers...."
```suggestion
Setting the number of consumers for an AMQP configuration and the Listener source determines how many consumers are spawned by the source and receive AMQP messages concurrently, each through its own channel. +
```
Is that correct?

> The message throughput relates to the number of consumers, and it enhances if the number of consumers increases, but it is also conditioned to the performance characteristics of the flow.

Reviewer: ```suggestion
Message throughput depends on the number of consumers, becoming greater as the number of consumers increases. However, throughput is also affected by the performance characteristics of the flow.
```
Such as?

> ===== Ack Mode

Reviewer: ```suggestion
===== ACK Mode Parameter
```

> You can set acknowledgment mode (ACK) at the configuration level (Consumer tab). Listener and Consume operations can also override this mode. ACK mode determines the acknowledgment mode to use when consuming from the AMQP broker. There are three acknowledgment modes: +

Reviewer: ```suggestion
ACK mode determines the acknowledgment mode to use when consuming from the AMQP broker. You set acknowledgment mode (ACK) at the configuration level (using the *Consumer* tab), but Listener and consume operations can override this mode. +
There are three acknowledgment modes: +
```

* `AUTO` +

> Use Automatic to acknowledge a received message automatically only if the flow execution finishes successfully.

Reviewer: ```suggestion
Use Automatic mode to acknowledge a received message automatically only if the flow execution finishes successfully.
```

* `MANUAL` +

> Use Manual if you want the app logic to control message acknowledgment.

Reviewer: ```suggestion
Use Manual mode if you want app logic to control message acknowledgment.
```

* `IMMEDIATE` +

> Use Immediate when you want to acknowledge messages as your app consumes them, before processing them.

Reviewer: ```suggestion
Use Immediate mode to acknowledge messages as your app consumes them, before processing them.
```

> Note that if a transaction is opened in the channel, the ACK mode parameter is overruled and the commit handles automatically the message acknowledgment, also the rollback handles automatically the recovery.

Reviewer: Is this in addition to how Listener and consume operations can override the ACK value (which we do not explain above but only mention)?
```suggestion
Note that if a transaction is opened in the channel instead of from the AMQP broker, the commit automatically acknowledges the message and the rollback automatically performs recovery.
```
Should this sentence be moved to Line 203?

> ===== Quality of Service Tab

Reviewer: ```suggestion
===== Quality of Service Prefetch Parameters
```
Should be parallel with ACK Mode peer title.

> You can set the following parameters at the configuration level (Quality of Service tab) with 0 as unlimited. Listener operation can also override the setting. These are the parameters:

Reviewer: ```suggestion
You set QoS parameters at the configuration level (using the *Quality of Service* tab), with 0 representing an unlimited value. A Listener operation can override this setting. +
These are the parameters:
```

* Prefetch Size +

> Defines a prefetch size window. The broker sends as many messages as possible without exceeding the prefetch size window in octets (bytes).

Reviewer: Is this what we really mean:
```suggestion
Defines the size of the prefetch window, in octets (bytes). The broker sends as many messages as possible without exceeding this size.
```

* Prefetch Count +

> Specifies a global prefetch window in terms of the max number of whole messages. This field may be used in combination with the `prefetch-size` field; a message will only be sent in advance if both prefetch windows allow it.

Reviewer: Similarly to the previous, does this really mean...:
```suggestion
Specifies the size of a prefetch window in terms of maximum number of whole messages. You can use this parameter in combination with the `prefetch-size` parameter to send a message in advance only it both prefetch windows sizes allow it.
```

> ===== Ack Mode and Prefetch

Reviewer: How is this parallel with its peers? The existing structure is...:

===AMQP Connector
====Consume Parameters
=====Number of Consumers Parameter
=====Acknowledgement Mode Parameter
=====Quality of Service Prefetch Parameters
...and then a topic that repeats two of the previous, 
=====ACK Mode and Prefetch?

Isn't the info in this fourth topic, a Concept topic, required to understand the previous two Reference topics about ACK and QoS?
Please consider in a future revision.
Following text copyedited only.

> Acknowledgment mode and QoS prefetch value have a significant effect on consumer throughput. In general, increasing prefetch improves the rate of message delivery to consumers. The Immediate acknowledgment mode yields the best possible rate of delivery. However, in both cases the number of delivered but not yet processed messages will also increase, therefore increasing consumer RAM consumption.

Reviewer: ```suggestion
The acknowledgment mode and QoS prefetch value that you choose have a significant effect on consumer throughput. In general, increasing the prefetch value and choosing Immediate acknowledgment mode yield the best possible rate of delivery. However, with both choices the number of delivered but not yet processed messages also increases, thereby increasing RAM consumption.
```

> Automatic acknowledgement mode or Manual acknowledgment mode with unlimited prefetch should be used with care. Consumers that consume a lot of messages without acknowledging will lead to memory consumption growth on the broker node they are connected to. Finding a suitable prefetch value is a matter of trial and error and can vary from workload to workload.

Reviewer: ```suggestion
Be very careful in choosing Automatic acknowledgement mode, or Manual acknowledgment mode paired with the unlimited prefetch value 0. Consumption of a lot of messages without acknowledgement leads to increased memory consumption on the broker node. Finding a suitable prefetch value is a matter of trial and error and can vary from workload to workload.
```
Shouldn't this say that finding a suitable ACK mode and prefetch value is a matter..., and should this statement actually start the topic?

==== Publish Parameters

> You can set the following parameters at the configuration level (Publisher tab). Publish and Publish-consume operations can also override these parameters. If the use case requires it, you can enable only one of the recommendations:

Reviewer: ```suggestion
Set the following parameters at the configuration level, using the *Publisher* tab. Publish and Publish-consume operations can override these parameters. Depending on your use case, you might need to enable only one of these parameters.
```

> ===== Delivery Mode

Reviewer: ```suggestion
===== Delivery Mode Parameter
```

> It declares the delivery mode to use when publishing to the AMQP broker. Possible parameters values are:

Reviewer: ```suggestion
Setting this parameter determines the delivery mode used to publish to the AMQP broker. Possible parameters values are:
```

* `PERSISTENT` +

> Messages marked as persistent and delivered to durable queues are logged to disk on the broker.

Reviewer: ```suggestion
Messages delivered to durable queues are logged to disk on the broker.
```

* `TRANSIENT` +

> No logging for the delivered message is performed.

Reviewer: ```suggestion
Delivered messages are not logged.
```

> The broker is responsible for the implementation of the recordable storage mechanism for persistent messages. The impact of this feature on performance depends on the AMQP server.

Reviewer: Shouldn't the first sentence be part of the `PERSISTENT` list item?
And does the second sentence also refer to `PERSISTENT`, or does it apply to both list items?
Clarify by positioning.

> ===== Request Broker Confirm

Reviewer: ```suggestion
===== Request Broker Confirm Parameter
```

> In case there is no confirmation provided by the broker, it dictates whether the operation must fail. Adds one extra communication between the broker and the client per message published.

Reviewer: ```suggestion
If the broker provides no confirmation, setting this parameter determines whether the operation fails. Using this parameter adds one extra communication between the broker and the client per message published.
```
What are the parameter settings: true and false? yes and no? some value vs. no value?
How does this add a layer of complexity, and what does that mean: increased overhead?
